neuron: cleanup neuron package for upstreaming (WIP)

pramodk opened the pull request:
~- [x] This is still WIP but outlined what "I think" relevant for upstream vs. what is our BBP/developer specific changes.~
~- [x] See my comments as "TODO" and all the code between "=============".~
~- [x] In order to avoid two completely different version in upstream and BBP, I suggest we discuss today the marked sections: @iomaganaris  @matz-e  @ohm314~ 

Here is the updated description:
- [x] NEURON CI for this branch is passing in https://github.com/neuronsimulator/nrn/pull/2636
- [x] @iomaganaris helped to fix the coreneuron perf update CI https://bbpgitlab.epfl.ch/hpc/cellular/coreneuron-perf-tests/-/pipelines/182574
- [x] blueconfig CI is also green now: https://bbpgitlab.epfl.ch/hpc/sim/blueconfigs/-/pipelines/182573

Changes done in this PR:
* remove duplicate neuron recipe in builtin vs patches repo
   * keep BBP dev-specific options in patches
   * the recipe that should go upstream is in the builtin repo  
* remove various legacy logic and variants in the recipe
* just keep the versions that would make sense (especially all alphas)
* some modernisation/refactorings based on the reviews

Commit:
neuron: cleanup neuron package for upstreaming (WIP)

this is still WIP but outlined what "I think" relevant for upstream
vs. what is our BBP/developer specific changes.

See my comments as "TODO" and all the code between "=============".

In order to avoid two completely different version in upstream and BBP,
I suggest we discuss the marked sections.

ohm314 commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
+
+    # add build type 'FastDebug' for result validations
+    # TODO: FastDebug should be removed? Maybe useful for internal tests
+    #       and not BBP CI setup. So should be OK to keep?
```
Since the default is reasonable this won't hurt to just stay as an option.

iomaganaris replied:
I think that this could be helpful to keep in the BBP package only for the ease of use when we have to build neurodamus-packages for reproducing issues, CI, etc

ohm314 commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
+
+    # sanitizer to help during development
+    # TODO: Should be removed as it's really only few of core developers
+    #       use it and no need as a general option?
```
agree

ohm314 commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
+    # extended set of model tests
+    # TODO: OK to keep as these can be used by others as generic
+    #       performance tests. But there is still a question in
+    #       my mind if/how anyone else is going to use it and how.
```
i would remove this. I think for the special case where you want to run performance tests you'd want to maybe build on the side, doesn't seem like a thing I'd want in a pkg manager.

iomaganaris replied:
I agree that this shouldn't be in the upstream package. For the BBP CI cases I think this helps a lot to simplify the CIs

ohm314 commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
         )
 
+        # simd vectorisation flag with gcc and nmodl code generation
+        # TODO: Should go into CMake IMO
```
agree

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
 
     homepage = "https://www.neuron.yale.edu/"
-    url = "http://www.neuron.yale.edu/ftp/neuron/versions/v7.5/nrn-7.5.tar.gz"
+    url = "https://github.com/neuronsimulator/nrn/releases/download/8.2.3/full-src-package-8.2.3.tar.gz"
```
This is great! I download this and then wildly guess later which software it actually is when I look at `~/Downloads`?

pramodk replied:
I thought the same when I copied the URL 🙄 Agree, we need to fix.

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
+    # dependencies for performance measurement
+    depends_on("caliper", when="+caliper")
+
+    # some common conflicts that use can build
+    conflicts("+rx3d", when="~python")
+    conflicts("+gpu", when="~coreneuron")
+    conflicts("+openmp", when="~coreneuron")
```
These seem moot with conditional variants.

pramodk replied:
that's being me not up to date! will do!

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
-    # ==============================================
-    # ==== CMake build system related functions ====
-    # ==============================================
+    # patches that are necessary for common build issues
```
What are these issues? Maybe just remove that comment.

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -21,434 +20,288 @@ class Neuron(CMakePackage):
-            cmake_name = spec_requiremement[1:].upper().replace("-", "_")
+        spec = self.spec
+
+        def cmake_options(spec_options):
```
`define_from_variant`! https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake

matz-e reviewed (dismissed):
This file should not contain the string `"-D`

iomaganaris reviewed (commented):
Some comments from quickly looking through the discussion items

pramodk committed:
Addressed various issues and discussions.

Main point is to have builtin package and bluebrain-repo
package that extends the builtin package.

pramodk committed:
push some changes

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -66,180 +48,123 @@ class Neuron(CMakePackage):
+            nmodl_options += " sympy --analytic"
+        if spec.satisfies("+sympyopt"):
+            nmodl_options += " --conductance --pade --cse"
+        args.append("-DCORENRN_NMODL_FLAGS=%s" % nmodl_options)
```
`self.define`...

pramodk replied:
@matz-e : could you check if I have updated all `-D..` correctly with `self.define` or equivalent?

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -66,180 +48,123 @@ class Neuron(CMakePackage):
+            # which do not run on exclusive resources and do not give reliable
+            # results for tests that test performance scaling
+            args.append(self.define("BUILD_TESTING", True))
+            args.append(self.define("NRN_ENABLE_PERFORMANCE_TESTS", False))
```
`define_from_variant` or similar

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -66,180 +48,123 @@ class Neuron(CMakePackage):
+        specific directory by looking for a specific binary.
+        """
+        return (
+            subprocess.Popen(
```
I think this could also be rewritten using `which("awk")(...)`

pramodk committed:
remove \+binary from neurodamus recipes

pramodk committed:
add commit that is used with the FastDebug flag in CMake

iomaganaris committed:
Add "develop" version in packages not having it in configure-pipeline

iomaganaris committed:
Small fixes

iomaganaris committed:
Fix search_file_with_regex() paramaters

iomaganaris committed:
Added comment in changes in configure_pipeline

pramodk committed:
remove versioning constraints to avoid preference for <9 neuron version

pramodk committed:
replace -Dfoo with self.define usage

pramodk committed:
add missing handling of reports

pramodk committed:
add shared variant back

pramodk committed:
make style check happy

pramodk committed:
add backtrace variant: incorporate BlueBrain/spack/pull/2117/

pramodk committed:
make flake happy

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -12,44 +12,88 @@ class Neuron(CMakePackage):
-    version("8.0.0", tag="8.0.0", submodules="True")
-    version("7.8.2", tag="7.8.2", submodules="True")
-    version("7.8.1", tag="7.8.1", submodules="True")
+    version("develop", branch="master")
```
Upstream will tell you this should be version `"master"` last I recall.

pramodk replied:
yeah, I remember and it's a bit annoying to have that changed. I guess there will be some more changes from the upstream review. After that, I pull the upstream recipe into our fork.

matz-e replied:
I'd move this to our `patch` repo.

pramodk replied:
done!

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -59,38 +103,64 @@ def cmake_args(self):
+            return self.define("NRN_ENABLE_" + cmake_name, value)
 
         args = [
             cmake_options(variant)
             for variant in [
+                "+backtrace",
                 "+coreneuron",
                 "+interviews",
                 "+mpi",
                 "+python",
                 "+rx3d",
-                "+coreneuron",
+                "+shared",
                 "+tests",
             ]
         ]
```
This is all `define_from_variant` :)

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -112,10 +182,10 @@ def filter_compilers(self):
 
         kwargs = {"backup": False, "string": True}
-        nrnmech_makefile = join_path(self.prefix, "./bin/nrnmech_makefile")
+        nrnmech_makefile = join_path(self.prefix, "bin/nrnmech_makefile")
```
I think both file filters here could be merged, too.

pramodk replied:
do you meant something like:

```console
-        if spec.satisfies("@:7"):
-            assign_operator = "?="
-        else:
-            assign_operator = "="
-
-        filter_file(
-            "CC {0} {1}".format(assign_operator, env["CC"]),
-            "CC = {0}".format(cc_compiler),
-            nrnmech_makefile,
-            **kwargs,
-        )
-
-        filter_file(
-            "CXX {0} {1}".format(assign_operator, env["CXX"]),
-            "CXX = {0}".format(cxx_compiler),
-            nrnmech_makefile,
-            **kwargs,
-        )
+        assign_operator = "?=" if spec.satisfies("@:7") else "="
+
+        # older neuron versions also use CC
+        compilers = [("CC", "cc_compiler"), ("CXX", "cxx_compiler")]
+
+        for compiler_var, compiler_env in compilers:
+            pattern = "{0} {1} {2}".format(
+                compiler_var, assign_operator,  env[compiler_env]
+            )
+            replacement = "{0} = {1}".format(
+                compiler_var, locals()[compiler_env]
+            )
+            filter_file(pattern, replacement, nrnmech_makefile, **kwargs)
```

I am also trying to not add too many changes/improvements to avoid breaking something (this makefile replacement on cray vs non-cray platforms was buggy)

matz-e replied:
Yes. as an added bonus, in the new deployment, no lines match `env[compiler_env]`, and I just matched `^CXX {assign_operator} .*` to get things working "better". Still broken, though. Upstream `neuron` will be pained by the same.

pramodk replied:
Done

pramodk committed:
fix TRUE vs ON in variant+cmake options

pramodk committed:
Address review: remove subprocess.Popen

pramodk committed:
minor refactpring

pramodk commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -1,488 +1,158 @@
-    version("develop", branch="master")
+    # BBP specific version
+    # TODO: 9.0.a18 for testing PR
+    version("9.0.a18", commit="a3b35127ea119b159394851c1b6bcf5f49c04784")
```
⚠️ : I will replace this commit after neuronsimulator/nrn/pull/2636 is merged!

iomaganaris reviewed (dismissed):
As far as I can see LGTM 👍 
Very welcome improvement!

pramodk committed:
add new neuron version 9.0.a15

pramodk committed:
handle FastDebug as we might use build_type=FastDebug with previous neuron versions <= 9.0.a14

pramodk commented:
The above pipeline failed because I manually canceled the running job so that we could launch a fresh build one more time before merging.

pramodk committed:
shared variant is mostly for coreneuron

heerener commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -1,488 +1,178 @@
-    depends_on("boost", when="@8.99:+tests+coreneuron")
-    depends_on("cuda", when="@8.99:+gpu")
-    depends_on("libsonata-report", when="@8.99:+report+coreneuron")
+    # used furing development by core team
```
`s/furing/during/`
But please do this in a future PR, don't hold back this PR for a nitpicky non-issue!

pramodk replied:
thanks for spotting that! 😒 

I will see if I need to push another commit otherwise I will merge it.

A bit painful thing has been that neuron-check CI is failing "randomly" only with GPU builds. If I retrigger that on different  GPU nodes (especially non-ldu* / phase2) then everything is passing).

I am checking this by explicitly launching neuron CI on this branch: https://github.com/neuronsimulator/nrn/pull/2646.

If that will be green and if re-triggered CI is green than I will merge this.

pramodk replied:
addressed as I need to fix something anyway.

heerener reviewed (dismissed):
Looks a lot more digestible than before!

Please also make sure to get an approval from the spack-master himself :-)

pramodk committed:
keep neuron at a14 and fix the typo

Commit:
py-neurodamus: bump to 2.17.0

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -1,488 +1,177 @@
-            ]
-        ]
-        if self.spec.satisfies("+tests"):
+        args = super(Neuron, self).cmake_args()
```
```suggestion
        args = super().cmake_args()
```
Python 3 only by now.

pramodk replied:
done in ff51c40c6b187e9c98e804b1bfa3c3a58e1af6a8

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -1,488 +1,177 @@
+        # FastDebug build type
+        if spec.satisfies("@:9.0.a14") and spec.variants["build_type"].value == "FastDebug":
+            # moderate optimisation by default
+            compilation_flags = ["-g", "-O1"]
             if "%intel" in self.spec:
-                # This one definitely seems wise
+                # this one definitely seems wise
                 compilation_flags += ["-fp-model", "consistent"]
             elif "%oneapi" in self.spec:
-                # The documentation doesn't mention consistent for these intel compilers
+                # the documentation doesn't mention consistent for these intel compilers
                 compilation_flags.append("-fp-model=precise")
```
We can't handle floating point handling in CMake?

pramodk replied:
note that this is for `@:9.0.a14` and the CMake change to handle such flags was added only recently. In neurodamus we often need to install older neuron versions for result cross-checking and hence I thought it would be helpful to have this.

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -1,488 +1,177 @@
+            args.append(self.define("CMAKE_C_FLAGS", compilation_flags))
+            args.append(self.define("CMAKE_CXX_FLAGS", compilation_flags))
+            # remove default flags (RelWithDebInfo etc.)
+            args.append("-DCMAKE_BUILD_TYPE=Custom")
```
Still a `-D` :)

pramodk replied:
coffee-pasta 😒. Done!

matz-e commented on bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:
Do we need the changes here, still? Could we not define the bogus `develop` version only in our package and then skip building upon the card house created by this file?

pramodk replied:
A valid point seeing this diff! I have no background of this `cmd/configure_pipeline.py` and wondering how you allowed to build this `card house` in the first place 😄 

@iomaganaris added this to fix coreneuron perf CI. I have to differ to him as I also don't know all the aspect about the perf-measurement CI.

iomaganaris replied:
The problem I tried to address here was an issue with the `configure-pipeline` command due to the fact that `version("develop")` is not defined in the patch recipe of NEURON since it's defined in the `builtin` recipe . I thought I would make more sense to add a `develop` version automatically if this command is used since there might be more recipes written without a `develop` version defined for some reason (either because it's a recipe built on top of a `builtin` one or there is `version("master"` instead). If you think it's not a good idea we could also define a `develop` version in the `patch` recipe of NEURON but there is a small chance we come across this issue again.

Edit: I saw you moved the `version("develop")` to the patch recipe so this fix shouldn't be needed anymore however I would propose throwing an error in case `develop` is not found in the recipe so we don't just continue building stuff with the incorrect branches, etc. (Could be done in a separate PR as well)

pramodk replied:
thanks for the explanation, Ioannis! (By tagging, didn't mean to bother you! we revisit this after the break!)

matz-e replied:
I just checked, and our new spack has a fix for installing `neuron@git.<sha>=develop`. This should make parsing recipes and modifying them a moot point once we merge the next deployment iteration.

Until then, no further complications to this script, please :)

Commit:
Revert "py-neurodamus: bump to 2.17.0"

This reverts commit 4953ed470b775a76f5e939fce0fd6c9cc2c12065.

pramodk committed:
Address review comments and make @matz-e Happy!

pramodk committed:
fix filter file and missing review changeset

pramodk reviewed (commented):
Addressed last comments from @matz-e. I will defer the aspect about `/cmd/configure_pipeline.py` as I don't have sufficient background. Otherwise everything should be in place!

matz-e commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -1,488 +1,177 @@
-                + ",".join(model for model in self.spec.variants["model_tests"].value)
+                self.define(
+                    "NRN_ENABLE_MODEL_TESTS",
+                    ",".join(model for model in spec.variants["model_tests"].value),
```
This can't be simplified?

pramodk replied:
Do you mean avoiding multi-valued variant altogether? that's possible but then we will remove flexibility and compile all the tests every time (which adds build time).

I would suggest to keep as it is unless @iomaganaris has different idea.

matz-e commented on bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:
I hope these modifications are not needed any longer?

matz-e reviewed (commented):
Looks good, last open point for me is that the `configure_pipeline` should really not need modifications (we should replace that if at all)

pramodk committed:
Revert changed to configure_pipeline

pramodk committed:
Merge branch 'develop' into pramodk/neuron-update

pramodk committed:
FastDebug build_type should be in BBP patches only (used for our CI)

pramodk committed:
Avoid regex matching with -DMPICH_SKIP_MPICXX=1

pramodk committed:
update doc string - same as upstream

iomaganaris commented on bluebrain/repo-patches/packages/neuron/package.py:
```diff
@@ -1,488 +1,185 @@
-# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
Nitpick
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

pramodk replied:
@iomaganaris: 2023 -> 2024 changes would be part of upstreaming merge I believe. So I will keep it as it is.

iomaganaris commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
Nitpick
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```
